backend/courses/management/commands/trainrecommender.py

```diff
@@ -0,0 +1,278 @@
+                                                                                                      courses_vectorized_by_description,
+                                                                                                      copresence_vectors,
+                                                                                                      copresence_vectors_past):
+        if use_descriptions:
+            if np.linalg.norm(description_vector) == 0:
+                continue
+            total_vector_curr = np.concatenate([schedule_vector, description_vector, copresence_vector * 2])
+            total_vector_past = np.concatenate([schedule_vector, description_vector, copresence_vector_past * 2])
+        else:
+            total_vector_curr = np.concatenate([schedule_vector, copresence_vector * 2])
+            total_vector_past = np.concatenate([schedule_vector, copresence_vector_past * 2])
+        courses_to_vectors_curr[course] = total_vector_curr / np.linalg.norm(total_vector_curr)
+        courses_to_vectors_past[course] = total_vector_past / np.linalg.norm(total_vector_past)
+    return courses_to_vectors_curr, courses_to_vectors_past
+
+
+def normalize_class_name(class_name):
+    """
+    Take in a class name and return the standard name for that class
+    """
+    course_obj: Course = lookup_course(class_name)
+    if course_obj is None:
+        return class_name
+    class_name = str(course_obj.primary_listing)
+    if " " in class_name:
+        class_name = class_name.split(" ")[0]
+    return class_name
+
+
+def cosine_similarity(vec_a, vec_b):
+    np.dot(vec_a, vec_b) / (np.linalg.norm(vec_a) * np.linalg.norm(vec_b))
+
+
+def generate_course_clusters(n_per_cluster=100):
+    """
+    Clusters courses and also returns a vector representation of each class (one for having taken that class now,
+    and another for having taken it in the past)
+    """
+    course_vectors_dict_curr, course_vectors_dict_past = generate_course_vectors_dict()
+    _courses, _course_vectors = zip(*course_vectors_dict_curr.items())
+    courses, course_vectors = list(_courses), np.array(list(_course_vectors))
+    num_clusters = round(len(courses) / n_per_cluster)
+    model = KMeans(n_clusters=num_clusters)
+    raw_cluster_result = model.fit_predict(course_vectors)
+    clusters = [[] for _ in range(num_clusters)]
+    for course_index, cluster_index in enumerate(raw_cluster_result):
+        clusters[cluster_index].append(courses[course_index])
+
+    cluster_centroids = [sum(course_vectors_dict_curr[course] for course in cluster) / len(cluster) for cluster in
+                         clusters]
+    return cluster_centroids, clusters, course_vectors_dict_curr, course_vectors_dict_past
+
+
+def save_course_clusters(cluster_data):
+    return pickle.dump(cluster_data, open("./course-cluster-data.pkl", "wb"))
```

Comment:
We'll have to figure out a different way to do this in production, because the local filesystem isn't persistent -- any time we deploy new code, we're in a new docker container with a clean filesystem. We could dump the pickle into redis? That's longer-lived. @kirubarajan @pawalt would love your input here.

Reply:
I'm partial to s3. A cronjob that dumps into s3 would make the most sense in my mind.

Reply:
I agree with @pawalt, s3 would be a pretty clean solution.

Reply:
I'm probably late here but S3 would be the industry standard for something like this.

Reply:
I'm thinking S3 and Redis are the optimal solution here (check out the most recent version of the PR)